### DISCUSSION TITLE
📝 Formatter causes `ts-ignore` and `ts-expect-error` to be misplaced

### DISCUSSION THREAD

**Discussion by nico-w** [biomejs/biome#3086] [Ideas]
Hey there,

I think this is not really a "bug", so let's start a discussion on it 😊

When we ran the biome formatter on our project the very first time, it was causing a couple of issues regarding TS directives that we placed in our code.

Code will pass the TS checks before formatting, but after formatting with e.g. 120 line width configured, the `ts-expect-error` will be misplaced and TS check won't pass any longer. We manually needed to check a couple of files and depending on the size of the project, this could take a while, hindering people from easily migrating.

While working on changes, this might not be a big difficulty but doing smart replaces in the code base overall can also cause this to happen.

Example snippet:
```
type MyFunction = (
  myVeryFirstArgument: string,
  mySecondArgument: number,
) => void

const myVeryLongDescriptiveFunctionNameThatDoesSomeStuffWithArguments: MyFunction =
  (myVeryFirstArgument, mySecondArgument) => {
    console.log('myVeryFirstArgument', myVeryFirstArgument)
    console.log('mySecondArgument', mySecondArgument)
  }

// @ts-expect-error
myVeryLongDescriptiveFunctionNameThatDoesSomeStuffWithArguments('this is a valid string, correct type','this is also a string, but invalid type')
```
See here: [Playground-Link](https://biomejs.dev/playground/?lineWidth=120&code=dAB5AHAAZQAgAE0AeQBGAHUAbgBjAHQAaQBvAG4AIAA9ACAAKAAKACAAIABtAHkAVgBlAHIAeQBGAGkAcgBzAHQAQQByAGcAdQBtAGUAbgB0ADoAIABzAHQAcgBpAG4AZwAsAAoAIAAgAG0AeQBTAGUAYwBvAG4AZABBAHIAZwB1AG0AZQBuAHQAOgAgAG4AdQBtAGIAZQByACwACgApACAAPQA%2BACAAdgBvAGkAZAAKAAoAYwBvAG4AcwB0ACAAbQB5AFYAZQByAHkATABvAG4AZwBEAGUAcwBjAHIAaQBwAHQAaQB2AGUARgB1AG4AYwB0AGkAbwBuAE4AYQBtAGUAVABoAGEAdABEAG8AZQBzAFMAbwBtAGUAUwB0AHUAZgBmAFcAaQB0AGgAQQByAGcAdQBtAGUAbgB0AHMAOgAgAE0AeQBGAHUAbgBjAHQAaQBvAG4AIAA9AAoAIAAgACgAbQB5AFYAZQByAHkARgBpAHIAcwB0AEEAcgBnAHUAbQBlAG4AdAAsACAAbQB5AFMAZQBjAG8AbgBkAEEAcgBnAHUAbQBlAG4AdAApACAAPQA%2BACAAewAKACAAIAAgACAAYwBvAG4AcwBvAGwAZQAuAGwAbwBnACgAJwBtAHkAVgBlAHIAeQBGAGkAcgBzAHQAQQByAGcAdQBtAGUAbgB0ACcALAAgAG0AeQBWAGUAcgB5AEYAaQByAHMAdABBAHIAZwB1AG0AZQBuAHQAKQAKACAAIAAgACAAYwBvAG4AcwBvAGwAZQAuAGwAbwBnACgAJwBtAHkAUwBlAGMAbwBuAGQAQQByAGcAdQBtAGUAbgB0ACcALAAgAG0AeQBTAGUAYwBvAG4AZABBAHIAZwB1AG0AZQBuAHQAKQAKACAAIAB9AAoACgAvAC8AIABAAHQAcwAtAGUAeABwAGUAYwB0AC0AZQByAHIAbwByAAoAbQB5AFYAZQByAHkATABvAG4AZwBEAGUAcwBjAHIAaQBwAHQAaQB2AGUARgB1AG4AYwB0AGkAbwBuAE4AYQBtAGUAVABoAGEAdABEAG8AZQBzAFMAbwBtAGUAUwB0AHUAZgBmAFcAaQB0AGgAQQByAGcAdQBtAGUAbgB0AHMAKAAnAHQAaABpAHMAIABpAHMAIABhACAAdgBhAGwAaQBkACAAcwB0AHIAaQBuAGcALAAgAGMAbwByAHIAZQBjAHQAIAB0AHkAcABlACcALAAnAHQAaABpAHMAIABpAHMAIABhAGwAcwBvACAAYQAgAHMAdAByAGkAbgBnACwAIABiAHUAdAAgAGkAbgB2AGEAbABpAGQAIAB0AHkAcABlACcAKQA%3D)

What do you think?

**Comment by ematipico** [comment]
Unfortunately, the formatter doesn't know that that particular comment is special. There isn't much we can do for now. You can increase the line width to avoid collapsing or manually move the comment.

**Reply by arendjr** [reply]
Do you think we should be open to a PR that gives the formatter such knowledge? I imagine it might be tricky to get right, but it would solve a user issue. I know even the “manual moving” suggestion can be annoying because you move it and then on format it jumps again, so you really have to anticipate how the formatter is going to intervene.

**Reply by ematipico** [reply]
I am not against expanding the semantics of the comments. However, it can't be quickly done as a fix. There are repercussions in terms of semantics and expectations for the users. It's definitely a new feature we can add

**Reply by Conaclos** [reply]
Prettier has the same issue? I wonder if they plan to fix it?

**Reply by ematipico** [reply]
> Prettier has the same issue?

Yes, you can check the playground. 

>  I wonder if they plan to fix it?

I don't know
